src/peergos/server/storage/GarbageCollector.java

```diff
@@ -0,0 +1,146 @@
+package peergos.server.storage;
+
+import peergos.server.util.*;
+import peergos.shared.cbor.*;
+import peergos.shared.crypto.hash.*;
+import peergos.shared.io.ipfs.multihash.*;
+import peergos.shared.storage.*;
+
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.*;
+import java.util.logging.*;
+
+public class GarbageCollector implements ContentAddressedStorage {
+
+    private static final long MAX_WAIT_FOR_TRANSACTION_MILLIS = 10_000;
+
+    private final ContentAddressedStorage target;
+    private final long gcPeriodMillis;
+    private final Object gcLock = new Object();
+    private final ConcurrentHashMap<PublicKeyHash, AtomicInteger> openTransactions = new ConcurrentHashMap<>();
+
+    public GarbageCollector(ContentAddressedStorage target, long gcPeriodMillis) {
+        this.target = target;
+        this.gcPeriodMillis = gcPeriodMillis;
+    }
+
+    public void start() {
+        new Thread(this::run).start();
+    }
+
+    private int openTransactions() {
+        int res = 0;
+        for (AtomicInteger open : openTransactions.values()) {
+            res += Math.max(0, open.get());
+        }
+        return res;
+    }
+
+    public void run() {
+        while (true) {
```

Comment:
Can you add a way to signal this runnable to finish?

Reply:
It doesn't need to be cleanly shutdown